proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+| Field | Type | Description |
+|-------|------|-------------|
+| `count` | `varuint32` | The number of types in the signature |
+| `type` | `value_type*` | The type of each element in the signature |
+
+
+##### external_kind
+
+A single-byte unsigned integer indicating the kind of definition being imported
+or defined:
+
+* `0` indicating a `Function` [import](Modules.md#imports) or [definition](Modules.md#function-and-code-sections)
+* `1` indicating a `Table` [import](Modules.md#imports) or [definition](Modules.md#table-section)
+* `2` indicating a `Memory` [import](Modules.md#imports) or [definition](Modules.md#linear-memory-section)
+* `3` indicating a `Global` [import](Modules.md#imports) or [definition](Modules.md#global-section)
+* `4` indicating an `Exception` [import](#import-section) or [definition](#exception-sectio)
+
+### Module structure
+
+#### High-level structure
+
+A new `exception` section is introduced and is named `exception`. If included,
+it must appear between the `Export` and `Start` sections of the module.
+
+
+##### Exception section
+
+The `exception` section is the named section 'exception'. The exception section
+declares exception types using exception type signatures.
+
+| Field | Type | Description |
+|-------|------|-------------|
+| count | `varuint32` | count of the number of exceptions to follow |
+| sig | `except_type*` | The type signature of the data fields for each exception |
+
+
+##### Import section
+
+The import section is extended to include exception types by extending an
+`import_entry` as follows:
+
+If the `kind` is `Exception`:
+
+| Field | Type | Description |
+|-------|------|-------------|
+| `sig`  | `except_type` | the type signature of the exception |
+
+##### Export section
+
+The export section is extended to include exception types by extending an
+`export_entry` as follows:
+
+If the `kind` is `Exception`, then the `index` is into the corresponding
+exception in the [exception index space](#exception-index-space).
+
+
+##### Name section
+
+The set of known values for `name_type` of a name section is extended as
+follows:
+
+| Name Type | Code | Description |
+| --------- | ---- | ----------- |
+| [Function](#function-names) | `1` | Assigns names to functions |
+| [Local](#local-names) | `2` | Assigns names to locals in functions |
+| [Exception](#exception-names) | `3` | Assigns names to exception types |
+
+###### Exception names
+
+The exception names subsection is a `name_map` which assigns names to a subset
+of the _exception_ indices from the exception section. (Used for both imports
+and module-defined).
+
+### Control flow operators
+
+The control flow operators are extended to define try blocks, catch blocks,
+throws, and rethrows as follows:
+
+| Name | Opcode | Immediates | Description |
+| ---- | ---- | ---- | ---- |
+| `try` | `0x06` | sig : `block_type` | begins a block which can handle thrown exceptions |
+| `catch` | `0x07` | | begins the catch block of the try block |
```

Comment:
So this is supposed to be `catch_all`? If we are gonna add static the `catch tag` instruction later, shouldn't this be `catch_all` to make them different?

Reply:
Agreed with @aheejin. OTOH, we may not even need to distinguish the two. Instead, catch has a list of exception ids that defines what it catches. In the degenerate case that it is empty it is a catch-all. In the MVP, we could then require it to be empty (although I don't necessarily think we should).

Reply:
So to make sure I understand the idea, you mean have something like:
```
try:
  ...
catch 0, 1:
  ...
catch 2:
  ...
catch ε:
  ...
```

I kind of like that idea (although we'd probably want to require disjoint exception indices for each catch clause, and aliasing might be a problem).

Or did you mean have a single catch clause overall?

The reason for allowing only catch all for MVP is that our main consumer, C++, doesn't really have any use for non-catch_all. I think it makes sense to not spec and implement a feature until we have users for it. Are there other languages that are interested in being among the first exception users?

Reply:
Yes, I meant still just a single exception clause, but annotated with an (optional) list of exception ids. That should give us the best of both worlds: simple instruction structure and efficient implementation.

Why does C++ not have use for specific catch? I would assume that any catch that is not a `catch(...)` in C++ could be using this. Or does LLVM make that impossible? Either way, other languages would probably only catch specific exceptions most of the time.

Reply:
I think we'd still want to be able to associate different code with different catch clauses. For example, say my language didn't follow the one exception type per language convention, but instead had Wasm exceptions for things like `FileNotFoundException`, `NetworkTimeoutException`, etc. I'd want to be able to handle these differently. Although there is a workaround of just nesting try blocks, or dynamically inspecting the exception inside the catch clause.

The issue with C++ and specific catch was discussed some in #31. @aheejin can probably explain the issues better but I think it comes down to a combination of LLVM's exception handling being very unstructured, and needing `catch(...)` to apply to both C++ exception and JavaScript exceptions. Basically, Clang/LLVM generates a big chunk of code that checks if an exception is known to C++ or not and does the appropriate thing. It sounds like to use specific catch we'd basically have to duplicate the exception handling code in the catch_all branch, and even finding which code to duplicate is non-trivial.

Reply:
Differentiating clauses only really makes sense if you also unpack the payload data, because the code will probably want to use that (and if you have to do that separately anyway then you can as well distinguish code paths there) -- which would bring us back to the earlier design without a separate match instruction. I'm fine with that as well, but the separation has the appeal that it simplifies instruction structure and is a better fit for the problems caused by LLVM.

For C++, I assume there isn't much win in compiling `catch(E e)` more efficiently because the handler stack is littered with implicit catch-all anyway, because destructors? Because if it wasn't then I wouldn't see why you couldn't use `catch $C++exn` to implement `catch(E e)` more efficiently -- that isn't concerned with foreign exceptions.

Reply:
> Differentiating clauses only really makes sense if you also unpack the payload data, because the code will probably want to use that (and if you have to do that separately anyway then you can as well distinguish code paths there) -- which would bring us back to the earlier design without a separate match instruction.

Ah, right. I had been assuming that the specific catch form would also unpack the exception data. It seems like this would let the embedder implement exception handling without materializing any kind of exception object if it is not used, which seems like a likely performance win.

> For C++, I assume there isn't much win in compiling `catch(E e)` more efficiently because the handler stack is littered with implicit catch-all anyway, because destructors? Because if it wasn't then I wouldn't see why you couldn't use `catch $C++exn` to implement `catch(E e)` more efficiently -- that isn't concerned with foreign exceptions.

Yes, pretty much everything gets surrounded by a catch-all to run destructors. The problem with `catch(E e)` is that it also catches any subclass of `E`, meaning libcxxabi has to look at the RTTI to find out whether it can handle an exception.

Reply:
I agree that a combined catch/match makes certain optimisations easier, but I'm not sure how big a deal it is to do a data flow analysis instead. OTOH, the separation allows for more flexible control flow, which I thought was the motivation?

Subtyping on C++ exns seems like an orthogonal issue -- they'd still all have the same signle exn tag, don't they? But I can see how the destructor business makes this mostly a moot point in C++.

Reply:
The goal is to get both. Keep the dynamic exception matching for cases where you need to more flexible control flow, but allow the more efficient static version when it's sufficient. I'm not sure how valuable this is though, and the complexity is significantly higher to support both in the spec. As you point out, a dataflow analysis would work, and I expect this would be a local analysis making it cheap to do (you wouldn't have to examine the whole program before you start compiling).

I see what you mean about C++ exceptions. If there are no destructors, then there's no reason to handle JavaScript exceptions at all unless there's a `catch(...)`. I'm pretty sure there are basically always destructors though, probably even more so if you're using exceptions.

Reply:
I think the idea of `catch` instruction that takes variable number of tags (or indices), as @rossberg suggested, haven't been discussed yet. Is it possible for an instruction to take variable number of immediate values?

Reply:
I can't think of any instructions that currently do this, but there's nothing fundamentally impossible about it. Rather than encoding it as `catch exception_index`, we'd just encode it as `catch vec(exception_index)`, similar to how [function types](https://webassembly.github.io/spec/core/binary/types.html#function-types) are encoded.

Reply:
See `br_table`, which has  vec(labelidx) as immediate.